csharp/src/SeedLang.Ast/Executor.cs

```diff
@@ -0,0 +1,78 @@
+// Copyright 2021 The Aha001 Team.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+
+namespace SeedLang.Ast {
+  // A executor class to execute an program represented by an AST tree.
+  //
+  // The executor traverses through the AST tree by implementing the interfaces of the AstVisitor
+  // class. The visit method of an expression will return the result of the expression. The return
+  // value of a statement visit method is not used.
+  public sealed class Executor : AstVisitor<BaseValue> {
+    // The global environment of the executor.
+    private readonly Dictionary<string, BaseValue> _globals = new Dictionary<string, BaseValue>();
+
+    public Executor() {
+    }
+
+    // Executes the given AST tree.
+    public BaseValue Run(AstNode node) {
+      return Visit(node);
+    }
+
+    // Registers a native function with a name in the global environment.
+    public void RegisterNativeFunc(string name, Func<BaseValue, BaseValue> func) {
+      // TODO: handle the ArgumentException when the name already exists.
+      _globals.Add(name, new NativeFuncValue(func));
+    }
+
+    protected internal override BaseValue VisitBinaryExpression(BinaryExpression binary) {
+      BaseValue left = Visit(binary.Left);
+      BaseValue right = Visit(binary.Right);
+      // TODO: handle other operators.
+      switch (binary.Op) {
+        case BinaryOperator.Add:
+          return left + right;
+        case BinaryOperator.Subtract:
+          return left - right;
+        case BinaryOperator.Multiply:
+          return left * right;
+        case BinaryOperator.Divide:
+          return left / right;
+        default:
+          throw new ArgumentException("Unsupported binary operator.");
+      }
+    }
+
+    protected internal override BaseValue VisitNumberConstant(NumberConstantExpression number) {
+      return new NumberValue(number.Value);
+    }
+
+    protected internal override BaseValue VisitStringConstant(StringConstantExpression str) {
+      return new StringValue(str.Value);
+    }
+
+    protected internal override BaseValue VisitEvalStatement(EvalStatement eval) {
+      if (_globals["print"] is NativeFuncValue func) {
```

Comment:
Is it the best choice to connect the built-in "print" function with the eval statement directly like this?

The semantic meaning of the eval statement is just "it results in a value" or "it returns a value", instead of "it prints a value to the standard output".

In a stack-based VM, a statement that returns a value simply leaves the value on the top of the stack. After that, the system may decide whether and how to use that value - a shell or a calculator may print the value out, while an interpreter may simply ignore the value and proceed to the next statement. Does a register-based VM share a similar behavior?

How about defining a consistent method to hold the return value of the eval statement somewhere in the VM? Hence, the calculator APP will have enough flexibility to make its own decision on how to run an executor and how to handle the result of an eval statement.

Reply:
Remove the print native function and expose the result through a visualizer.